src/CodeGen_OpenCL_Dev.cpp

```diff
@@ -209,8 +211,39 @@ string simt_intrinsic(const string &name) {
 void CodeGen_OpenCL_Dev::CodeGen_OpenCL_C::visit(const For *loop) {
     user_assert(loop->for_type != ForType::GPULane)
         << "The OpenCL backend does not support the gpu_lanes() scheduling directive.";
+       
+    // Cover the added GPUThreadReduceLoop
+    if(loop->for_type == ForType::GPUThreadReduce) {
+        // Check if the condition for the gpu_thread_reduce holds, only for addition (right now)
+        user_assert(loop->body.as<Store>()->value.as<Add>())
+            << "GPUThreadReduce is only defined on addition\n";
+
+        // get the local ID of this tread
+        stream << get_indent() << print_type(Int(32)) << " " << print_name(loop->name)
+               << " = " << simt_intrinsic(loop->name) << ";\n";
 
-    if (is_gpu_var(loop->name)) {
+        // initialize the repsentive sum value (might already be loaded in case of sum[0])
+        stream << get_indent() << "_sum[" << print_name(loop->name) << "] = _sum[" << print_name(loop->name) << "] + "<< print_name(loop->name) << ";\n";
+
+        // Wait for all threads to do this
+        stream << get_indent() << "barrier(CLK_LOCAL_MEM_FENCE);\n";
```

Comment:
To be exact I think you need a CLK_GLOBAL_MEM_FENCE as well since you are storing the results in global memory (sum)

Reply:
However if you want to use local memory, this is fine as is